src/System.Runtime.Numerics/src/System/Numerics/BigCalc.SquMul.cs

```diff
@@ -0,0 +1,356 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigCalc
+    {
+#if DEBUG
+
+        private const int SquareThreshold = 128 / 32;
+        private const int MultiplyThreshold = 128 / 32;
+
+#else
+
+        private const int SquareThreshold = 1024 / 32;
+        private const int MultiplyThreshold = 1024 / 32;
+
+#endif
+
+        public static uint[] Square(uint[] value)
+        {
+            Debug.Assert(value != null);
+
+            return Square(value, value.Length);
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static uint[] Square(uint[] value, int valueLength)
```

Comment:
I would propose just setting a local variable `valueLength = value.Length`, which doesn't hurt performance at all (because you already do it in the caller) and makes the method signature cleaner. I understand caching the length of the array, but shouldn't we do it internally rather than expose it externally?